File: ios/HybridRtmpPublisherView+DisconnectCause.swift
//
//  HybridRtmpPublisherView+DisconnectCause.swift
//  NitroRtmpPublisher
//
//  Best-effort disconnect-cause classifier. iOS surfaces a socket drop only as a
//  generic RTMP code (`NetConnection.Connect.Closed`) — identical whether the
//  phone lost its uplink, a call grabbed the audio session, or the ingest server
//  hung up. This module correlates side-channels captured at drop time to label
//  the cause in the `.disconnect` / `.connectionfailed` / `.reconnecting` message
//  (no API change — the existing `(event, message)` callback just gets a more
//  specific message, e.g. `network-lost (NetConnection.Connect.Closed)`):
//
//    • NWPathMonitor            → network-lost / network-changed(iface); ALSO
//                                  fires the proactive stop+restart when the
//                                  default path leaves the session's interface
//                                  (see restartStreamForNetworkChange below —
//                                  that part is behavior, not just labeling)
//    • audio-session .began     → interrupted-by-call (phone call / Siri / alarm)
//    • ProcessInfo thermalState → thermal-critical
//    • UIDevice batteryLevel    → low-battery (best effort only — a true battery
//                                  death kills the process before any callback
//                                  can fire, so this catches at most the last gasp)
//    • app background           → app-backgrounded
//    • none of the above        → server-closed (link's up, no local cause)
//
//  Everything here is ADVISORY: a mislabel never changes reconnect behavior, it
//  only annotates the message, so the cross-thread reads are deliberately cheap.
//

import Foundation
import Network
import UIKit

extension HybridRtmpPublisherView {

  // ─── NWPathMonitor lifecycle ─────────────────────────────────────────────

  /// Start watching the network path. Called once from `init`. The handler runs
  /// on `networkMonitorQueue`; it stashes the latest reachability + interface
  /// behind `netStateLock` so `disconnectCauseCategory` (RTMP status thread) can
  /// read it race-free. Also stamps `_lastNetworkLossUptime` on every drop so a
  /// brief flap (a Wi-Fi→cellular handoff that re-satisfies within a few seconds)
  /// is still attributable when the socket death lands just after re-connection.
  /// Additionally fires the PROACTIVE network-change restart when the default
  /// path's interface moves away from the one the live session is bound to.
  func startNetworkMonitor() {
    guard networkMonitor == nil else { return }
    let monitor = NWPathMonitor()
    monitor.pathUpdateHandler = { [weak self] path in
      guard let self else { return }
      let satisfied = (path.status == .satisfied)
      let iface: String
      if path.usesInterfaceType(.wifi) { iface = "wifi" }
      else if path.usesInterfaceType(.cellular) { iface = "cellular" }
      else if path.usesInterfaceType(.wiredEthernet) { iface = "ethernet" }
      else if path.usesInterfaceType(.other) { iface = "other" }
      else { iface = "none" }
      self.netStateLock.lock()
      if !satisfied { self._lastNetworkLossUptime = ProcessInfo.processInfo.systemUptime }
      self._networkSatisfied = satisfied
      self._networkInterface = iface
      let sessionIface = self._sessionInterface
      self.netStateLock.unlock()
      // Proactive network-change restart (Android parity:
      // onDefaultNetworkAvailable → restartStreamForNetworkChange). The default
      // path now runs on a DIFFERENT interface than the one the session bound
      // at connect time. This matters most for cellular→wifi joins: iOS keeps
      // established TCP flows on cellular (make-before-break), so without this
      // the stream would silently stay on cellular — burning mobile data with
      // wifi available and never emitting any event. The shouldBeStreaming read
      // here is only a cheap pre-filter; restartStreamForNetworkChange re-checks
      // everything serialized on main.
      if satisfied, isRealInterface(sessionIface), isRealInterface(iface),
         iface != sessionIface, self.shouldBeStreaming {
        self.restartStreamForNetworkChange(reason: "network-changed(\(iface))")
      }
    }
    networkMonitor = monitor
    monitor.start(queue: networkMonitorQueue)
  }

  func stopNetworkMonitor() {
    networkMonitor?.cancel()
    networkMonitor = nil
  }

  // ─── Battery monitoring ──────────────────────────────────────────────────

  /// Enable battery telemetry so the classifier has a level/state to read. Must
  /// run on main (UIKit); cheap and idempotent. The cached scalars are refreshed
  /// here and by the `batteryStateDidChange` observers registered in `init`.
  func enableBatteryMonitoring() {
    onMain { [weak self] in
      UIDevice.current.isBatteryMonitoringEnabled = true
      self?.refreshBatteryCache()
    }
  }

  /// Notification target (main thread) for battery level/state changes. Also
  /// invoked once from `enableBatteryMonitoring` to prime the cache.
  @objc func batteryStateDidChange() {
    refreshBatteryCache()
  }

  /// Read UIDevice battery on main and cache the scalars the classifier reads.
  /// Reading UIDevice off-main trips the Main Thread Checker, hence the cache.
  func refreshBatteryCache() {
    cachedBatteryLevel = UIDevice.current.batteryLevel
    cachedBatteryUnplugged = (UIDevice.current.batteryState == .unplugged)
  }

  // ─── Classifier ──────────────────────────────────────────────────────────

  /// Map a raw RTMP close / `connectFailed` code to a likely human cause,
  /// e.g. `"network-lost (NetConnection.Connect.Closed)"`. The bare category
  /// comes from `disconnectCauseCategory`; the raw code is appended for
  /// diagnostics so nothing is lost.
  func classifyDisconnectCause(rtmpCode: String) -> String {
    return "\(disconnectCauseCategory()) (\(rtmpCode))"
  }

  /// The bare cause category, ordered most-specific/decisive first so a single
  /// dominant signal wins. Shared by the socket-close path and the silent-stall
  /// watchdog. See the file header for what each signal means.
  func disconnectCauseCategory() -> String {
    // 1) App suspended — iOS reaps the RTMP socket on background.
    if isInBackground { return "app-backgrounded" }

    // 2) Phone call / Siri / alarm — an audio-session interruption is live, or
    //    fired within the last few seconds, or the capture session is paused by
    //    the audio device being grabbed by another client. A definitive local
    //    signal, so it outranks a network read (the call is WHY the link idled).
    let now = ProcessInfo.processInfo.systemUptime
    if audioInterruptionActive || captureInterrupted
        || (lastAudioInterruptionUptime > 0 && now - lastAudioInterruptionUptime < 8) {
      return "interrupted-by-call"
    }

    // 3) Uplink gone — path currently unsatisfied, or dropped within the last 5s
    //    (a Wi-Fi↔cellular handoff re-satisfies quickly but still kills the
    //    in-flight socket, so a bare "currently satisfied" check would miss it).
    netStateLock.lock()
    let satisfied = _networkSatisfied
    let iface = _networkInterface
    let lastLoss = _lastNetworkLossUptime
    let sessionIface = _sessionInterface
    netStateLock.unlock()
    if !satisfied { return "network-lost" }
    if lastLoss > 0 && now - lastLoss < 5 { return "network-changed(\(iface))" }
    // 3b) No unsatisfied blip was recorded, but the default path's interface no
    //     longer matches the one this session bound at (re)connect time — a
    //     make-before-break handoff (typically a cellular→wifi join): iOS brings
    //     the new path up before the old flow dies, so the socket death lands
    //     with the path "satisfied" throughout and the 5s window above misses
    //     it. Without this the drop would fall through to the server-closed
    //     guess. Android parity: the live-vs-bound handle check (3b) in
    //     +DisconnectCause.kt.
    if isRealInterface(sessionIface), isRealInterface(iface), iface != sessionIface {
      return "network-changed(\(iface))"
    }

    // 4) Thermal shutdown pressure (iOS throttles/kills the app when critical).
    if ProcessInfo.processInfo.thermalState == .critical { return "thermal-critical" }

    // 5) Battery critically low & unplugged — best effort: a real battery death
    //    kills us before any event fires, so this only ever catches the last gasp.
    if cachedBatteryLevel >= 0, cachedBatteryLevel < 0.05, cachedBatteryUnplugged {
      return "low-battery"
    }

    // 6) Link's up, no local interruption → the server hung up on us
    //    (idle timeout, key expiry, ingest restart, rate-limit, …).
    return "server-closed"
  }

  // ─── Network-change recovery ─────────────────────────────────────────────

  /// Stamp the interface the (re)starting session binds to — the baseline the
  /// proactive network-change trigger and the classifier's 3b check compare
  /// the live path against. Called from `startStream` and from each reconnect
  /// attempt (a reconnect often lands on a different network).
  func baselineSessionInterface() {
    netStateLock.lock()
    _sessionInterface = _networkInterface
    netStateLock.unlock()
  }

  /// Network-change recovery, kept dead simple by product decision (Android
  /// parity: +Reconnect.kt's restartStreamForNetworkChange): a plain public
  /// stopStream() + startStream() cycle — the sequence that reliably re-homes
  /// RTMP onto the new interface. stopStream() flips shouldBeStreaming, which
  /// gates off both the reactive reconnect machinery and duplicate path-update
  /// triggers, so the paths can't fight; the delayed startStream restores it.
  func restartStreamForNetworkChange(reason: String) {
    onMain { [weak self] in
      guard let self else { return }
      guard self.shouldBeStreaming, !self.isInBackground else { return }
      // Capture the URL BEFORE stopStream() (it nulls lastFullStreamUrl).
      guard let url = self.lastFullStreamUrl else { return }
      self.log("Recovery (\(reason)) → stop + restart stream")
      // Stamp the DISCONNECT stopStream() is about to emit with why we
      // dropped, so JS can tell a network switch from a user stop.
      self.pendingDisconnectReason = reason
      try? self.stopStream()
      // stopStream() cancelled any pending resume, so schedule ours AFTER it —
      // and a genuine USER stopStream() during the settle re-cancels this, so
      // a deliberate stop is never resurrected. The 1.5s settle (Android
      // parity: NETWORK_CHANGE_SETTLE_MS) lets the new interface finish coming
      // up before the fresh connect.
      let work = DispatchWorkItem { [weak self] in
        guard let self else { return }
        self.networkChangeResumeWork = nil
        // A user re-start (or anything else) claimed the session during the
        // settle — theirs wins; don't stack a second publish on it. And if the
        // app backgrounded meanwhile, a connect would only burn a timeout.
        guard !self.shouldBeStreaming, !self.isPublishingInFlight(),
              !self.isInBackground else { return }
        // Re-read the interface at fire time: the settle gave the OS time to
        // finish the switch, so the RECONNECTING that precedes the fresh start
        // names the network we're ACTUALLY resuming on (a double-flap during
        // the settle would otherwise carry a stale label).
        self.netStateLock.lock()
        let iface = self._networkInterface
        let satisfied = self._networkSatisfied
        self.netStateLock.unlock()
        let refined = (satisfied && isRealInterface(iface))
          ? "network-changed(\(iface))" : reason
        self.emitConnectionEvent(.reconnecting, refined)
        try? self.startStream(url: url)
      }
      self.networkChangeResumeWork = work
      DispatchQueue.main.asyncAfter(deadline: .now() + 1.5, execute: work)
    }
  }
}

/// True for a concrete session-bindable interface label — filters the
/// "no baseline yet" ("", "unknown") and pathless ("none") states out of the
/// network-change detection.
private func isRealInterface(_ iface: String) -> Bool {
  iface == "wifi" || iface == "cellular" || iface == "ethernet" || iface == "other"
}

File: android/src/main/java/com/margelo/nitro/rtmppublisher/HybridRtmpPublisherView+DisconnectCause.kt
package com.margelo.nitro.rtmppublisher

import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.media.AudioManager
import android.net.ConnectivityManager
import android.net.Network
import android.net.NetworkCapabilities
import android.os.BatteryManager
import android.os.Build
import android.os.PowerManager
import android.os.SystemClock

/**
 * Best-effort disconnect-cause classifier — Android mirror of
 * [HybridRtmpPublisherView+DisconnectCause.swift].
 *
 * RootEncoder surfaces a socket failure only as a raw reason string
 * (`onConnectionFailed(reason)`) or a bare `onDisconnect()` — identical whether
 * the phone lost its uplink, a call grabbed the mic, or the ingest server hung
 * up. This module correlates side-channels sampled at failure time to label the
 * cause in the emitted `onConnectionEvent` message (no API change — the message
 * string just gets more specific, e.g. `network-lost (broken pipe)`):
 *
 *  - [ConnectivityManager] default-network callback → network-lost /
 *    network-changed(iface)
 *  - [AudioManager.getMode]  → interrupted-by-call (cellular MODE_IN_CALL /
 *    VoIP MODE_IN_COMMUNICATION — no READ_PHONE_STATE permission required)
 *  - [PowerManager] thermal status → thermal-critical
 *  - sticky ACTION_BATTERY_CHANGED → low-battery (best effort only — a true
 *    battery death kills the process before any callback fires)
 *  - surface torn down → app-backgrounded; none of the above → server-closed
 *
 * Everything here is ADVISORY: a mislabel never changes reconnect behavior, it
 * only annotates the message, so the reads are deliberately cheap.
 */

// ─── ConnectivityManager lifecycle ──────────────────────────────────────────

internal fun HybridRtmpPublisherView.registerNetworkMonitor() {
  if (networkCallback != null) return
  val cm = connectivityManager ?: return
  val cb = object : ConnectivityManager.NetworkCallback() {
    override fun onAvailable(network: Network) {
      networkAvailable = true
      onDefaultNetworkAvailable(network)
    }
    override fun onLost(network: Network) {
      networkAvailable = false
      lastNetworkLossUptimeMs = SystemClock.elapsedRealtime()
    }
    override fun onCapabilitiesChanged(network: Network, caps: NetworkCapabilities) {
      networkTransport = transportLabelFor(caps)
      networkAvailable = true
    }
  }
  safe("registerNetworkMonitor") {
    cm.registerDefaultNetworkCallback(cb)
    networkCallback = cb
  }
}

internal fun HybridRtmpPublisherView.unregisterNetworkMonitor() {
  val cm = connectivityManager ?: return
  val cb = networkCallback ?: return
  safe("unregisterNetworkMonitor") { cm.unregisterNetworkCallback(cb) }
  networkCallback = null
}

// Called from the default-network callback's onAvailable. A new default
// network became available; if its handle differs from the one we were on AND
// we're mid-stream, the default network switched under us — the live RTMP
// socket is bound to the dead interface, so force a full rebuild bound to the
// new network. Fires PROACTIVELY (before Pedro even reports the socket break),
// and the reconnectInProgress latch dedups against the failure-driven path.
internal fun HybridRtmpPublisherView.onDefaultNetworkAvailable(network: Network) {
  val handle = network.networkHandle
  val prev = activeNetworkHandle
  // First network seen — baseline only; the normal connect path owns the
  // initial link (startStreamInternal re-baselines at every start anyway).
  if (prev == 0L) {
    activeNetworkHandle = handle
    return
  }
  if (prev == handle) return
  // Default network switched under a live stream → stop + restart. Deliberately
  // do NOT overwrite activeNetworkHandle here: it tracks the network the CURRENT
  // session is bound to (the restart's startStreamInternal re-baselines it), and
  // both the disconnect classifier's 3b check and the restart's late re-labeling
  // rely on the pre-switch value to recognize "the default moved under us".
  // Transport is read live — the cached networkTransport lags the switch (the
  // new network's onCapabilitiesChanged hasn't landed yet), which used to label
  // a cellular→wifi join "network-changed(cellular)".
  restartStreamForNetworkChange("network-changed(${currentTransportLabel()})")
}

// ─── Classifier ─────────────────────────────────────────────────────────────

/**
 * Label a raw RootEncoder failure reason with a likely cause, e.g.
 * `"network-lost (broken pipe)"`. The bare category comes from
 * [disconnectCauseCategory]; a non-blank raw reason is appended for diagnostics.
 */
internal fun HybridRtmpPublisherView.classifyDisconnectCause(rawReason: String): String {
  val category = disconnectCauseCategory()
  return if (rawReason.isBlank()) category else "$category ($rawReason)"
}

/**
 * The bare cause category, ordered most-specific/decisive first so a single
 * dominant signal wins. Shared by the failure path and the silent-stall
 * watchdog. See the file header for what each signal means.
 */
internal fun HybridRtmpPublisherView.disconnectCauseCategory(): String {
  // 1) Surface gone = app backgrounded (Pedro's socket dies with the surface).
  //    Note: both terminal call-sites early-return while !surfaceReady, so this
  //    is a defensive fallback rather than a common path.
  if (!surfaceReady) return "app-backgrounded"

  // 2) Phone call / VoIP — read the audio mode (no permission needed). A
  //    definitive local signal, so it outranks a network read (the call is WHY
  //    the mic/link idled).
  val mode = audioManager?.mode
  if (mode == AudioManager.MODE_IN_CALL || mode == AudioManager.MODE_IN_COMMUNICATION) {
    return "interrupted-by-call"
  }

  // 3) Uplink gone — currently unavailable, or dropped within the last 5s (a
  //    Wi-Fi↔cellular handoff re-attaches quickly but still kills the in-flight
  //    socket, so a bare "currently available" check would miss it).
  val now = SystemClock.elapsedRealtime()
  if (!networkAvailable) return "network-lost"
  if (lastNetworkLossUptimeMs > 0 && now - lastNetworkLossUptimeMs < 5_000) {
    return "network-changed(${currentTransportLabel()})"
  }
  // 3b) The default network switched, but neither onLost nor the proactive
  //     onDefaultNetworkAvailable callback has updated our signals yet: both can
  //     land AFTER Pedro reports the dead socket, and a fresh interface coming up
  //     while the old one lingers fires no onLost at all (so the 5s window above
  //     misses it — this is the "network switch mislabeled server-closed" case).
  //     Compare the LIVE default-network handle against the one the current
  //     socket is bound to: a mismatch is a switch we haven't processed, so name
  //     it rather than falling through to the server-closed guess below. Reads
  //     the transport live (cached networkTransport also lags the switch).
  //     activeNetworkHandle==0L ⇒ never baselined ⇒ skip.
  val boundHandle = activeNetworkHandle
  if (boundHandle != 0L) {
    val liveHandle = connectivityManager?.activeNetwork?.networkHandle
    if (liveHandle != null && liveHandle != boundHandle) {
      return "network-changed(${currentTransportLabel()})"
    }
  }

  // 4) Thermal shutdown pressure.
  if (isThermalCritical()) return "thermal-critical"

  // 5) Battery critically low & unplugged — best effort: a real battery death
  //    kills us before any callback fires, so this catches at most the last gasp.
  if (isBatteryCriticallyLow()) return "low-battery"

  // 6) Link's up, no local cause → the server hung up on us (idle timeout, key
  //    expiry, ingest restart, rate-limit, …).
  return "server-closed"
}

// Map a NetworkCapabilities to our coarse transport label. Shared by the
// onCapabilitiesChanged cache and the live read below so the two can't drift.
private fun transportLabelFor(caps: NetworkCapabilities): String = when {
  caps.hasTransport(NetworkCapabilities.TRANSPORT_WIFI) -> "wifi"
  caps.hasTransport(NetworkCapabilities.TRANSPORT_CELLULAR) -> "cellular"
  caps.hasTransport(NetworkCapabilities.TRANSPORT_ETHERNET) -> "ethernet"
  else -> "other"
}

// Transport of the CURRENT default network, read straight from the system so it
// reflects a just-completed switch even before our onCapabilitiesChanged lands.
// Falls back to the cached [networkTransport] on any read hiccup. internal: also
// used by restartStreamForNetworkChange's late re-labeling in +Reconnect.kt.
internal fun HybridRtmpPublisherView.currentTransportLabel(): String =
  safe("currentTransportLabel", default = networkTransport) {
    val net = connectivityManager?.activeNetwork ?: return@safe networkTransport
    val caps = connectivityManager?.getNetworkCapabilities(net) ?: return@safe networkTransport
    transportLabelFor(caps)
  }

private fun HybridRtmpPublisherView.isThermalCritical(): Boolean {
  if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) return false
  // Prefer the cached value the thermal listener keeps fresh; otherwise a single
  // binder call. THERMAL_STATUS_CRITICAL and above means the OS is actively
  // throttling / about to shut components down.
  val status = if (thermalListenerRegistered) {
    lastThermalStatusInt
  } else {
    safe("thermalStatusForCause", default = PowerManager.THERMAL_STATUS_NONE) {
      powerManager?.currentThermalStatus ?: PowerManager.THERMAL_STATUS_NONE
    }
  }
  return status >= PowerManager.THERMAL_STATUS_CRITICAL
}

private fun HybridRtmpPublisherView.isBatteryCriticallyLow(): Boolean {
  return safe("batteryForCause", default = false) {
    // Sticky broadcast → returns the last battery state synchronously, no
    // receiver registration and no permission.
    val intent: Intent = context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))
      ?: return@safe false
    val level = intent.getIntExtra(BatteryManager.EXTRA_LEVEL, -1)
    val scale = intent.getIntExtra(BatteryManager.EXTRA_SCALE, -1)
    val statusInt = intent.getIntExtra(BatteryManager.EXTRA_STATUS, -1)
    if (level < 0 || scale <= 0) return@safe false
    val pct = level.toFloat() / scale.toFloat()
    val charging = statusInt == BatteryManager.BATTERY_STATUS_CHARGING ||
      statusInt == BatteryManager.BATTERY_STATUS_FULL
    pct < 0.05f && !charging
  }
}
